src/mame/machine/eispc_kb.cpp

```diff
@@ -0,0 +1,446 @@
+	PORT_BIT( 0x0100, IP_ACTIVE_LOW, IPT_KEYBOARD ) PORT_NAME("F8")       PORT_CODE(KEYCODE_F8)         PORT_CHAR(UCHAR_MAMEKEY(F8))                 PCM(key, 5) // 66
+	PORT_BIT( 0x0200, IP_ACTIVE_LOW, IPT_KEYBOARD )                       PORT_CODE(KEYCODE_LSHIFT) PORT_CHAR(UCHAR_SHIFT_1)                         PCM(key, 5) // 42
+	PORT_BIT( 0x0400, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_Z)          PORT_CHAR('z') PORT_CHAR('Z')                  PCM(key, 5) // 44
+	PORT_BIT( 0x0800, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_M)          PORT_CHAR('m') PORT_CHAR('M')                  PCM(key, 5) // 50
+	PORT_BIT( 0x1000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_A)          PORT_CHAR('a') PORT_CHAR('A')                  PCM(key, 5) // 30
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_S)          PORT_CHAR('s') PORT_CHAR('S')                  PCM(key, 5) // 31
+	PORT_BIT( 0x4000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_B)          PORT_CHAR('b') PORT_CHAR('B')                  PCM(key, 5) // 48
+	PORT_BIT( 0x8000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_N)          PORT_CHAR('n') PORT_CHAR('N')                  PCM(key, 5) // 49
+INPUT_PORTS_END
+
+
+//-------------------------------------------------
+//  ROM( eispc_kb )
+//-------------------------------------------------
+
+ROM_START( eispc_kb )
+	ROM_REGION( 0x800, M6801_TAG, 0 )
+	ROM_LOAD( "sgtpepper-1.2.bin", 0x000, 0x800, CRC(7107b841) SHA1(a939dd50622575c31fea9c7adb7a7db5403a7aca) )
+ROM_END
+
+} // anonymous namespace
+
+//**************************************************************************
+//  DEVICE DEFINITIONS
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(EISPC_KB, eispc_keyboard_device, "eispc_kb", "Ericsson PC keyboard")
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  eispc_keyboard_device - constructor
+//-------------------------------------------------
+
+eispc_keyboard_device::eispc_keyboard_device(
+		machine_config const &mconfig,
+		char const *tag,
+		device_t *owner,
+		uint32_t clock)
+	: device_t(mconfig, EISPC_KB, tag, owner, clock)
+	, m_mcu(*this, M6801_TAG)
+	, m_rows(*this, "P1%u", 0)
+	, m_txd_cb(*this)
+	, m_rxd_high(true)
+	, m_txd_high(true)
+	, m_hold(true)
+	, m_col_select(0)
+{
+}
+
+INPUT_CHANGED_MEMBER( eispc_keyboard_device::key )
+{
+		if (oldval && !newval)
+	{
+		LOGUI("Key Pressed - name: %s field: %04x param: %04x oldval: %04x newval: %04x\n", field.name(), field.defvalue(), param, oldval, newval);
+		int idx = *((int *)(&param));
+		if (idx >= sizeof(keys))
```

Comment:
Use `ARRAY_LENGTH` to get number of elements.

Reply:
Neat